src/protobuf.zig

```diff
@@ -449,21 +533,38 @@ fn append(pb: *ArrayList(u8), comptime field: FieldDescriptor, value: anytype, c
 /// Internal function that decodes the descriptor information and struct fields
 /// before passing them to the append function
 fn internal_pb_encode(pb: *ArrayList(u8), data: anytype) Allocator.Error!void {
-    const field_list = @typeInfo(@TypeOf(data)).Struct.fields;
-    const data_type = @TypeOf(data);
+    const type_info_data = @typeInfo(@TypeOf(data));
+    const data_type = switch (type_info_data) {
+        .Union => |_| // ManagedStruct case
+        @typeInfo(@TypeOf(data.Borrowed)).Pointer.child,
+        else => @TypeOf(data),
+    };
+    const field_list = @typeInfo(data_type).Struct.fields;
 
     inline for (field_list) |field| {
-        switch (@typeInfo(field.type)) {
-            .Optional => {
-                if (@field(data, field.name)) |value| {
-                    try append(pb, @field(data_type._desc_table, field.name), value, true);
-                }
-            },
-            else => try append(pb, @field(data_type._desc_table, field.name), @field(data, field.name), false),
+        if (@typeInfo(field.type) == .Optional) {
+            const temp = if (isZigProtobufManagedStruct(@TypeOf(data)))
+                data.get()
+            else
+                data;
```

Comment:
Since it is a very repeating pattern, I think there is a trivial way to write this if as the suggestion below resulting in the same final compiled code.
```suggestion
            const temp = getPointerOrValue(data);
```